.github/workflows/linux_builds.yml

```diff
@@ -212,3 +213,57 @@ jobs:
         if: ${{ matrix.artifact }}
         with:
           name: ${{ matrix.cache-name }}
+
+      - name: Dump GDExtension API for compatibility checking
+        if: ${{ matrix.compat }}
+        run: |
+          ${{ matrix.bin }} --headless --dump-extension-api
+
+      - name: Upload extension API
+        uses: actions/upload-artifact@v3
+        if: ${{ matrix.compat }}
+        with:
+          name: extension-api
+          path: extension_api.json
+          # Use the default artifact retention of 90 days to ensure the compat checking
+          # step below works as long as possible for PRs based on old commits
+
+      # Last step of this workflow as when merging a PR that intentionally breaks compat
+      # the failure of this step must not hide any other issues
+      - name: Check for GDExtension compatibility
+        if: ${{ matrix.compat }}
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          # download the reference extension_api.json from the previous build that compatibility should be checked against
+          get_api_ref() {
+            # fetch the git history so that we can get the previous commits
+            # but exclude diffs etc to keep the size down
+            git fetch --unshallow --filter=tree:0 > /dev/null 2> /dev/null
+            # For PRs check compatibility against the base commit, ie check the compatibility breaks from the entire PR if it has multiple commits
+            # For commits on branches check for compatibility against the previous commit
+            parent="${{ github.event.pull_request.base.sha }}"
+            if [ -z "$parent" ]; then parent="${{ github.sha }}^"; fi
+            # For the parent commit (and up to 100 parents, in case of batch merges or if the CI of some commits hasn't completed yet)
+            while read -r commit; do
+              # For each workflow id for the commit
+              while read -r workflow_id; do
+                # For each artifact named 'extension-api' of the workflow
+                while read -r download_url; do
+                  echo "Found the previous extension api dump of commit $commit."
+                  # Download and extract the artifact
+                  file=$(mktemp)
+                  gh api $download_url > "$file"
+                  unzip -o "$file" > /dev/null
+                  rm "$file"
+                  # And exit out after the first artifact has been found
+                  return 0
+                done < <(gh api -H "Accept: application/vnd.github+json" -H "X-GitHub-Api-Version: 2022-11-28" /repos/${{ github.repository }}/actions/runs/$workflow_id/artifacts --jq '.artifacts[] | select(.name == "extension-api") | .archive_download_url')
+              done < <(gh api -H "Accept: application/vnd.github+json" -H "X-GitHub-Api-Version: 2022-11-28" /repos/${{ github.repository }}/actions/runs?head_sha=$commit --jq '.workflow_runs[] | .id')
```

Comment:
Maybe instead of directly interfacing with the GitHub API via `gh api` this could use "action-download-artifact"?

I'm using this in a PR for godot-cpp to download the latest godot-editor-mono artifact from the 'master' branch, in order to run a test project with a GDExtension: https://github.com/godotengine/godot-cpp/pull/1101/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR134

I had to workaround a bug in the action, but you can certainly use my fork until the issue is fixed upstream :-)

Reply:
In my dev branch (I don't want to spam this PR with test commit too much) I already moved this all into a local action to make it reusable for https://github.com/godotengine/godot/pull/76647#discussion_r1181738997, so that cleans up the workflow a lot.

But the biggest issue is that the logic for which artifacts are acceptable for compat checking and which are preferred is fairly specific (we cannot risk using an artifact from a build that has commits that the current one does not, or we risk false positives with apis appearing to be removed when they just never existed in the PR) and that action does not appear to support a logic and AFAIK complex interactions between a script and an action are not really possible so we cannot just the commit selection logic.

Reply:
Ok, that makes sense! Looking forward to seeing the latest code :-)